# **PRE-PROCESSED RENT_CONTACTS**

In [7]:
import numpy as np
import pandas as pd

In [8]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [9]:
df_rent = pd.read_csv('/content/drive/MyDrive/dataset/AREA_TABLE1.csv')
df_rent.head()

,area_id,avg_rent,contract_count
0,230.0,25100.185100,11297
1,231.0,75793.572191,5938
2,232.0,199651.961889,397
3,233.0,22897.687917,48705
4,234.0,71716.020395,19891


In [10]:
print("Offices area in rent_contact")
df_rent["area_id"].nunique()

Offices area in rent_contact


168

# **PRE- PROCESSED TRANSACTION DATASET**

In [11]:
df_transaction = pd.read_csv('/content/drive/MyDrive/dataset/FINAL_TRANSACTION1.1.csv')
df_transaction.head()

,area_id,avg_sale_price,avg_sale_value,transaction_count,parking_score,metro_score,mall_score
0,229.0,1326.324301,2.029900e+06,551,0.000000,0.000000,0.000000
1,230.0,5143.045255,1.593337e+06,1473,0.000000,0.935506,0.000000
2,231.0,9686.377121,1.188933e+07,733,0.000000,0.961801,0.156889
3,232.0,7309.106103,2.817807e+06,9952,0.235129,0.971865,0.971865
4,233.0,9927.209538,2.645607e+06,2834,0.000000,0.982357,0.524700


# **PRE- PROCESSING THE EXCEL FILE**

In [12]:
import pandas as pd

df = pd.read_excel("survey_responses.xlsx")
df.to_csv("survey.csv", index=False)

In [13]:
survey = pd.read_csv("survey.csv")
survey.head()

,Timestamp,Feature 1 : Capital Appreciation Potential,Feature 2 : Off-Plan Speculative Momentum,Feature 3 : Commercial Liquidity & Deal Volume,Feature 4 : Building Parking Sufficiency,Feature 5 : Current Rental Yield Baseline,Feature 6 : Corporate Space Absorption Velocity,Feature 7 : District Supply Tightness Rate,Feature 8 : Transit-Oriented Accessibility Index,Feature 9 : Business Lifestyle Convenience Index,Feature 10 : Regulatory & Tax Shield Status,Feature 11 : Grade A Premium Asset Density,Feature 12 : Strategic Highway Bottleneck Access
0,2026-05-22 11:00:42.632,10,8,9,6,10,8,8,7,8,9,7,8


In [14]:
scores = survey.drop(columns=['Timestamp']).mean()  # mean over all survey responses

In [15]:
normalized = scores / scores.sum()
print(normalized)

Feature 1 : Capital Appreciation Potential           0.102041
Feature 2 : Off-Plan Speculative Momentum            0.081633
Feature 3 : Commercial Liquidity & Deal Volume       0.091837
Feature 4 : Building Parking Sufficiency             0.061224
Feature 5 : Current Rental Yield Baseline            0.102041
Feature 6 : Corporate Space Absorption Velocity      0.081633
Feature 7 : District Supply Tightness Rate           0.081633
Feature 8 : Transit-Oriented Accessibility Index     0.071429
Feature 9 : Business Lifestyle Convenience Index     0.081633
Feature 10 : Regulatory & Tax Shield Status          0.091837
Feature 11 : Grade A Premium Asset Density           0.071429
Feature 12 : Strategic Highway Bottleneck Access     0.081633
dtype: float64


# **JOINING THE 2 DATASETS (TRANSACTION, RENT_CONTACT)**

In [16]:
final_df = df_transaction.merge(
    df_rent,
    on = "area_id",
    how = "inner"
)

In [17]:
print(final_df.shape)
display(final_df.head())

(165, 9)


,area_id,avg_sale_price,avg_sale_value,transaction_count,parking_score,metro_score,mall_score,avg_rent,contract_count
0,230.0,5143.045255,1.593337e+06,1473,0.000000,0.935506,0.000000,25100.185100,11297
1,231.0,9686.377121,1.188933e+07,733,0.000000,0.961801,0.156889,75793.572191,5938
2,232.0,7309.106103,2.817807e+06,9952,0.235129,0.971865,0.971865,199651.961889,397
3,233.0,9927.209538,2.645607e+06,2834,0.000000,0.982357,0.524700,22897.687917,48705
4,234.0,7744.678114,6.808329e+06,599,0.000000,0.814691,0.716194,71716.020395,19891


In [18]:
final_df.shape

(165, 9)

In [19]:
 #Immediately check what you kept vs lost
print(f"Transaction districts:  {df_transaction['area_id'].nunique()}")
print(f"Rent districts:         {df_rent['area_id'].nunique()}")
print(f"After inner join:       {final_df['area_id'].nunique()}")
print(f"\nDistricts lost (only in transactions): "
      f"{set(df_transaction.area_id) - set(df_rent.area_id)}")
print(f"Districts lost (only in rent):         "
      f"{set(df_rent.area_id) - set(df_transaction.area_id)}")


# Sanity check — should be zero NaNs
print(f"\nNaN check: {final_df.isnull().sum().sum()} total NaNs")

Transaction districts:  259
Rent districts:         168
After inner join:       165

Districts lost (only in transactions): {513.0, 514.0, 524.0, 527.0, 528.0, 529.0, 229.0, 236.0, 241.0, 243.0, 245.0, 246.0, 250.0, 251.0, 252.0, 253.0, 256.0, 257.0, 258.0, 259.0, 262.0, 263.0, 265.0, 268.0, 270.0, 272.0, 273.0, 274.0, 275.0, 277.0, 279.0, 280.0, 281.0, 282.0, 283.0, 284.0, 285.0, 292.0, 298.0, 302.0, 327.0, 340.0, 349.0, 354.0, 377.0, 379.0, 380.0, 382.0, 383.0, 385.0, 386.0, 387.0, 392.0, 394.0, 403.0, 406.0, 411.0, 412.0, 413.0, 415.0, 417.0, 418.0, 419.0, 420.0, 421.0, 423.0, 424.0, 425.0, 427.0, 432.0, 434.0, 436.0, 448.0, 449.0, 451.0, 468.0, 469.0, 473.0, 476.0, 479.0, 480.0, 481.0, 486.0, 494.0, 496.0, 500.0, 501.0, 502.0, 503.0, 505.0, 507.0, 508.0, 509.0, 510.0}
Districts lost (only in rent):         {472.0, 346.0, 324.0}

NaN check: 0 total NaNs


In [20]:
common_area_ids = set(df_transaction['area_id']).intersection(set(df_rent['area_id']))
print(f"Number of common area_ids between current df_transaction and df_rent: {len(common_area_ids)}")
print(f"Common area_ids: {sorted(list(common_area_ids))}")

# This count should match final_df.shape[0], which is 17

Number of common area_ids between current df_transaction and df_rent: 165
Common area_ids: [230.0, 231.0, 232.0, 233.0, 234.0, 235.0, 237.0, 238.0, 239.0, 240.0, 242.0, 244.0, 247.0, 248.0, 249.0, 254.0, 255.0, 264.0, 266.0, 267.0, 269.0, 271.0, 276.0, 278.0, 288.0, 294.0, 295.0, 296.0, 297.0, 299.0, 300.0, 301.0, 303.0, 304.0, 305.0, 306.0, 307.0, 308.0, 309.0, 310.0, 311.0, 312.0, 313.0, 314.0, 315.0, 316.0, 317.0, 318.0, 319.0, 320.0, 321.0, 322.0, 323.0, 325.0, 326.0, 328.0, 329.0, 330.0, 331.0, 332.0, 333.0, 334.0, 335.0, 336.0, 337.0, 338.0, 339.0, 341.0, 342.0, 343.0, 344.0, 345.0, 347.0, 348.0, 350.0, 351.0, 352.0, 355.0, 356.0, 357.0, 358.0, 359.0, 360.0, 361.0, 362.0, 363.0, 364.0, 365.0, 366.0, 367.0, 368.0, 369.0, 370.0, 371.0, 372.0, 373.0, 374.0, 375.0, 376.0, 378.0, 384.0, 388.0, 390.0, 391.0, 393.0, 395.0, 396.0, 397.0, 398.0, 399.0, 400.0, 401.0, 402.0, 404.0, 405.0, 409.0, 410.0, 414.0, 416.0, 430.0, 431.0, 435.0, 437.0, 438.0, 441.0, 442.0, 443.0, 444.0, 445.0, 447.0

As you can see, the `area_id`s in the current `df_transaction` (ranging from 0 to 257) and `df_rent` (ranging from 230 to 531) have a smaller overlap. The intersection only contains `area_id`s from 230 to 246 (specifically: 230, 231, 232, 233, 234, 235, 237, 238, 239, 240, 242, 244, 246, 247, 248, 249, 255), totaling 17 unique IDs. This is why your `final_df` now has 17 rows, and the output from cell `p5g0MBa41_1Q` is from a prior run where the `df_transaction` dataset likely had a different set of `area_id`s resulting in a larger overlap.

In [21]:
final_df.head()

,area_id,avg_sale_price,avg_sale_value,transaction_count,parking_score,metro_score,mall_score,avg_rent,contract_count
0,230.0,5143.045255,1.593337e+06,1473,0.000000,0.935506,0.000000,25100.185100,11297
1,231.0,9686.377121,1.188933e+07,733,0.000000,0.961801,0.156889,75793.572191,5938
2,232.0,7309.106103,2.817807e+06,9952,0.235129,0.971865,0.971865,199651.961889,397
3,233.0,9927.209538,2.645607e+06,2834,0.000000,0.982357,0.524700,22897.687917,48705
4,234.0,7744.678114,6.808329e+06,599,0.000000,0.814691,0.716194,71716.020395,19891


Let's check the descriptive statistics for `avg_rent`, `contract_count`, `metro_score`, and `mall_score` to understand their distributions and identify any potential anomalies like negative values.

In [22]:
columns_to_check = ['avg_rent', 'contract_count']

for col in columns_to_check:
    print(f"\nDescriptive statistics for '{col}':")
    display(final_df[col].describe())

    # Check for negative values if applicable (e.g., for counts or scores that shouldn't be negative)
    if (final_df[col] < 0).any():
        negative_count = (final_df[col] < 0).sum()
        print(f"Number of negative values in '{col}': {negative_count}")
    else:
        print(f"No negative values found in '{col}'.")


Descriptive statistics for 'avg_rent':


,avg_rent
count,1.650000e+02
mean,3.622407e+05
std,9.175031e+05
min,1.000000e+03
25%,5.316676e+04
50%,9.848029e+04
75%,2.178019e+05
max,7.587092e+06


No negative values found in 'avg_rent'.

Descriptive statistics for 'contract_count':


,contract_count
count,165.000000
mean,8900.127273
std,17713.159925
min,1.000000
25%,69.000000
50%,1543.000000
75%,10109.000000
max,125406.000000


No negative values found in 'contract_count'.


----------------------------------------------------------

# **FEATURE ENGINEERING**

## Deriving rental_yield

In [23]:
# 1.  Then add your derived features
# Derived feature: rental_yield = annual office rent / total sale value (a proper ratio).
# avg_sale_value is the per-district mean of actual_worth on sales (from the transactions step).
# NOTE: divide by avg_sale_value, NOT avg_sale_price (price per m^2) -> that gave the ~488% bug.
final_df["rental_yield"] = final_df["avg_rent"] / final_df["avg_sale_value"]

# IQR-cap to remove proxy extremes (same outlier method used on the other features)
q1, q3 = final_df["rental_yield"].quantile([0.25, 0.75])
iqr = q3 - q1
lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
n_capped = ((final_df["rental_yield"] < lower) | (final_df["rental_yield"] > upper)).sum()
final_df["rental_yield"] = final_df["rental_yield"].clip(lower, upper)
print(f"rental_yield = avg_rent / avg_sale_value; IQR-capped to [{lower:.4f}, {upper:.4f}] for {n_capped} district(s)")


rental_yield = avg_rent / avg_sale_value; IQR-capped to [-0.0796, 0.1469] for 18 district(s)


In [24]:
final_df["rental_yield"].isnull().sum() # NO NULL VALUES PERFECT

np.int64(0)

In [25]:
final_df["rental_yield"]

,rental_yield
0,0.015753
1,0.006375
2,0.070854
3,0.008655
4,0.010534
...,...
160,0.022929
161,0.043466
162,0.002341
163,0.065671


## Build survey weights from `normalized`

In [26]:
import pandas as pd

# Re-define survey and normalized within this cell for standalone execution
df = pd.read_excel("survey_responses.xlsx")
df.to_csv("survey.csv", index=False)
survey = pd.read_csv("survey.csv")
scores = survey.drop(columns=['Timestamp']).mean()  # mean over all survey responses
normalized = scores / scores.sum()

# Map our 8 data columns to the correct survey feature weights
# using the 'normalized' Series already computed from the survey CSV
col_to_feature = {
    'avg_sale_price'   : 'Feature 1 : Capital Appreciation Potential ',
    'transaction_count': 'Feature 3 : Commercial Liquidity & Deal Volume ',
    'parking_score'    : 'Feature 4 : Building Parking Sufficiency ',
    'rental_yield'     : 'Feature 5 : Current Rental Yield Baseline ',
    'contract_count'   : 'Feature 6 : Corporate Space Absorption Velocity ',
    'avg_rent'         : 'Feature 7 : District Supply Tightness Rate ',
    'metro_score'      : 'Feature 8 : Transit-Oriented Accessibility Index ',
    'mall_score'       : 'Feature 9 : Business Lifestyle Convenience Index ',
}

# Pull the raw weights for our 8 columns
raw_weights = {col: normalized[feat] for col, feat in col_to_feature.items()}

# Renormalise so they sum to 1.0 (absorbs the missing features 2, 10, 11, 12)
total = sum(raw_weights.values())
feature_weights = {col: w / total for col, w in raw_weights.items()}

print('Adjusted weights (sum = 1.0):')
for col, w in feature_weights.items():
    print(f'  {col:<22}  {w:.6f}')
print(f"  {'TOTAL':<22}  {sum(feature_weights.values()):.6f}")

Adjusted weights (sum = 1.0):
  avg_sale_price          0.151515
  transaction_count       0.136364
  parking_score           0.090909
  rental_yield            0.151515
  contract_count          0.121212
  avg_rent                0.121212
  metro_score             0.106061
  mall_score              0.121212
  TOTAL                   1.000000


## **WEIGHTED COMPOSITE SCORE**

Raw columns are on completely different scales — `avg_sale_price` is in AED,
`transaction_count` can exceed 10,000, while `mall_score` is 0–5.
Without normalisation, large-magnitude columns dominate the score regardless of weight.
Min-Max scaling brings every feature to [0, 1] before applying weights.

In [27]:
from sklearn.preprocessing import MinMaxScaler

# Correct the feature_cols list to include only existing columns
feature_cols = [
    "avg_sale_price",
    "transaction_count",
    "parking_score",
    "rental_yield",
    "avg_rent",
    "contract_count",
    "metro_score",
    "mall_score",
]

# Min-Max normalise all feature columns
scaler = MinMaxScaler()
norm_col_names = [col + '_norm' for col in feature_cols]
final_df[norm_col_names] = scaler.fit_transform(final_df[feature_cols])

print('\nNormalised features — first 3 rows:')
display(final_df[['area_id'] + norm_col_names].head(3))


Normalised features — first 3 rows:


,area_id,avg_sale_price_norm,transaction_count_norm,parking_score_norm,rental_yield_norm,avg_rent_norm,contract_count_norm,metro_score_norm,mall_score_norm
0,230.0,0.051536,0.010776,0.000000,0.107139,0.003177,0.090076,0.935506,0.000000
1,231.0,0.097359,0.005359,0.000000,0.043304,0.009859,0.047343,0.961801,0.156889
2,232.0,0.073382,0.072849,0.235298,0.482191,0.026186,0.003158,0.971865,0.971865


## Compute weighted model
For each area:
```
hotspot_score = Σ (normalised_feature_i × weight_i)
```

In [28]:
# Weighted sum of normalised features
final_df['hotspot_score_raw'] = sum(
    final_df[col + '_norm'] * feature_weights[col]
    for col in feature_cols
)

# Scale to 0–100
score_min = final_df['hotspot_score_raw'].min()
score_max = final_df['hotspot_score_raw'].max()
final_df['hotspot_score'] = (
    (final_df['hotspot_score_raw'] - score_min) / (score_max - score_min)
) * 100

# Rank (1 = best hotspot)
final_df['rank'] = final_df['hotspot_score'].rank(ascending=False).astype(int)

result_df = final_df[['area_id', 'hotspot_score', 'rank']].sort_values('rank').reset_index(drop=True)

print('=== TOP 15 HOTSPOT AREAS ===')
print(result_df.head(15).to_string(index=False))

=== TOP 15 HOTSPOT AREAS ===
 area_id  hotspot_score  rank
   526.0     100.000000     1
   330.0      98.809473     2
   350.0      94.518321     3
   332.0      87.448979     4
   390.0      86.134346     5
   331.0      85.837680     6
   441.0      82.771946     7
   442.0      81.589678     8
   351.0      80.581545     9
   375.0      80.542912    10
   414.0      79.801111    11
   482.0      79.637411    12
   478.0      78.313265    13
   371.0      78.079293    14
   485.0      76.503312    15


In [29]:
final_df.to_csv('/content/drive/MyDrive/dataset/FINAL_DATASET.csv', index=False)